src/ophyd_async/core/signal_backend.py

```diff
@@ -45,3 +54,51 @@ async def get_setpoint(self) -> T:
     @abstractmethod
     def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
         """Observe changes to the current value, timestamp and severity"""
+
+
+if TYPE_CHECKING:
+    RuntimeEnum = Literal
+else:
+
+    class _RuntimeEnumMeta(type):
+        # Intentionally immutable class variable
+        __enum_classes_created: Dict[FrozenSet[str], Type["RuntimeEnum"]] = {}
+
+        def __str__(cls):
+            if hasattr(cls, "_choices"):
+                return f"RuntimeEnum{list(cls._choices.keys())}"
+            return "RuntimeEnum"
+
+        @property
+        def choices(cls) -> Tuple[str]:
+            return tuple(cls._choices.keys())
+
+        def __getitem__(cls, choices):
+            if isinstance(choices, str):
+                choices = (choices,)
+            else:
+                if not isinstance(choices, tuple) or not all(
+                    isinstance(c, str) for c in choices
+                ):
+                    raise TypeError(
+                        f"Choices must be a str or a tuple of str, not {choices}."
+                    )
+                if len(set(choices)) != len(choices):
+                    raise TypeError("Duplicate elements in runtime enum choices.")
+
+            choices_frozenset = frozenset(choices)
+
+            # If the enum has already been created, return it (ignoring order)
+            if choices_frozenset in _RuntimeEnumMeta.__enum_classes_created:
+                return _RuntimeEnumMeta.__enum_classes_created[choices_frozenset]
+
+            # Create a new enum subclass
+            class _RuntimeEnum(cls):
+                _choices = {choice: choice for choice in choices}
+
+            _RuntimeEnumMeta.__enum_classes_created[choices_frozenset] = _RuntimeEnum
+            return _RuntimeEnum
+
+    class RuntimeEnum(metaclass=_RuntimeEnumMeta):
```

Comment:
Then we probably want a typehint here
```suggestion
    class RuntimeEnum(metaclass=_RuntimeEnumMeta):
        choices: ClassVar[FrozenSet[str]]
```